Fix false positive for multi dep single use statements

Elrendio opened the pull request:
Resolves #91

This also resolves some false detections of `not_my_dep::my_dep_name::stuff_in_my_dep` 😊 I couldn't find a way to avoid detection of `futures` in with the single line regexp
```rust
pub use {
    async_trait::{mod1, dep2},
    not_futures::{
        futures::{futures_mod1, futures_mod2::{futures_mod21, futures_mod22}}
    },
    reqwest,
};
```

This update does come at a small performance downside (that might be resolved by not instantiating a regexp per dependency?). On our repo of ~500k lines of machete goes from ~45s user time to ~48s of user time.

bnjbvr commented on src/search_unused.rs:
```diff
@@ -77,25 +77,41 @@ fn make_line_regexp(name: &str) -> String {
+    //   2. Matches the crate's name with optional sub-modules
+    //   3. Matches modules after the usage of the crate's name
+    //
+    // In order to avoid false usage detection of `not_my_dep::my_dep` the regexp unsures that the
```
```suggestion
    // In order to avoid false usage detection of `not_my_dep::my_dep` the regexp ensures that the
```

bnjbvr commented on src/search_unused.rs:
```diff
@@ -77,25 +77,41 @@ fn make_line_regexp(name: &str) -> String {
+    // engine doesn't support recursion). Therefore, sub modules are authorized up to 4 levels
+    // deep.
+
+    let sub_modules_match = r#"(?:::\w+)*(?:::\*|\s+as\s+\w+|::\{(?:[^{}]*(?:\{(?:[^{}]*(?:\{(?:[^{}]*(?:\{[^{}]*\})?[^{}]*)*\})?[^{}]*)*\})?[^{}]*)*\})?"#;
```
It's already a nice comment, but can I ask for a little bit more please: could you take each part of this regular expression, and tie it to the comment? (Which part is 1. in your list, 2. in your list, etc. and feel free to add more comment sections if necessary)

Elrendio replied:
I've linked the different parts 😊

bnjbvr reviewed (changes requested):
Thanks! Just want a bit more comments, please.

Elrendio commented on src/search_unused.rs:
```diff
@@ -77,25 +77,41 @@ fn make_line_regexp(name: &str) -> String {
-    // that it finds, which *should* be the end of the use statement, but oh well.
-    // `(?i){name}(?-i)` makes the match against the crate's name case insensitive.
-    format!(r#"use \{{\s[^;]*(?i){name}(?-i)\s*as\s*[^;]*\}};"#)
+    // Breaking down this Terrible regular expression: tries to match uses of the crate's name in
+    // compound `use` statement accross multiple lines.
```
Hi @bnjbvr , it's not just wrapping I've also updated the content therefore I'm not sure what you expect

Elrendio commented:
Hi @bnjbvr ,

Sorry for the delay, I missed your comments 😞 I've addressed them and added a `.rustfmt.toml` that seems to mostly fit with your expectations in order for other contributors to follow your lint rules automatically. If you're not aligned I can remove it.

Have a nice day!

bnjbvr reviewed (approved):
Thanks! I'm not a huge fan of the custom `.rustfmt.toml`, so I'll get rid of it.

Thanks a lot for the comments, they make the regular expression slightly easier to understand (what a monster have we created here, now wondering if using a custom Rust syntax parser written with `nom` would be easier…).

I've tried it out on one of my favorite Rust projects `wasmtime`:

- wall clock time took a 18% slowdown hit, from 720ms -> 856ms
- user time took a 35% slowdown hit, from 2.7s -> 3.67s

While these are impressive in terms of relative numbers, since the tool is still running in less than a second on such a large code base, I consider it fine, especially since it's correcter (and what's the point of being fast if code isn't correct).

Merging, thanks a bunch 🙏
